docs/tasks/administer-cluster/kubeadm-upgrade-ha.md

```diff
@@ -0,0 +1,179 @@
+- Upgrades the control plane components or rollbacks if any of them fails to come up.
+- Applies the new `kube-dns` and `kube-proxy` manifests and enforces that all necessary RBAC rules are created.
+- Creates new certificate and key files of apiserver and backs up old files if they're about to expire in 180 days.
+
+Next, manually upgrade your Software Defined Network (SDN).
+
+Your Container Network Interface (CNI) provider may have its own upgrade instructions to follow. Check the [addons](/docs/concepts/cluster-administration/addons/) page to find your CNI provider and see if there are additional upgrade steps necessary.
+
+It has been observed that in some cases after having upgraded the primary master the upgrade on some of the secondary masters failed initially (timed out waiting for the restarted static pods to come up) but worked just fine when retried after a minute or two.
+
+## Upgrade base software packages
+
+At this point all the static pod manifests in your cluster, i.e. API Server, Controller Manager, Scheduler, Kube Proxy have been upgraded, however the base software, i.e. `kubelet`, `kubectl`, `kubeadm` installed on your nodes’ OS are still of the old version. For upgrading the base software packages on all nodes they will be taken out (`cordon` if master, `drain` if worker node) upgraded and then taken in (`uncordon`) again one by one. 
+
+First, the master nodes the master nodes will be upgraded (run `kubectl cordon` / `kubectl uncordon` from whatever machine you have set up to issue `kubectl` commands):
+
+```shell
+$ kubectl cordon <MASTER-NODE-FQDN>
+$ systemctl stop keepalived kubelet docker
+# use your distro's package manager, e.g. 'yum' on RH-based systems
+# for the versions stick to kubeadm's output (see above)
+$ yum install -y kubelet-<NEW-K8S-VERSION> kubectl-<NEW-K8S-VERSION> kubeadm-<NEW-K8S-VERSION> kubernetes-cni-<NEW-CNI-VERSION>
+$ systemctl start `echo keepalived kubelet docker | tac`
+$ kubectl uncordon <MASTER-NODE-FQDN>
+```
+
+In this example an _rpm_-based system is assumed and `yum` is used for installing the upgraded software. On _deb_-based systems it will be `apt-get update` and then `apt-get install <PACKAGE>=<NEW-K8S-VERSION>` for all packages.
+
+Because on the worker nodes there is no keepalived to stop and start and all processes need to be moved to other nodes cleanly before shutting anything down, `drain` is used instead of `cordon`. In this process all containers will get moved from one node to another at least once. If some of them keep local data you will need some strategy for this scenario. This really depends on the application(s) running in the cluster, and there is no simple, generic solution to this.
+
+On the master (or wherever you run `kubectl` commands):
+
+```shell
+$ kubectl drain --delete-local-data --ignore-daemonsets <WORKER-NODE-FQDN>
+```
+
+On the respective worker node:
+
+```shell
+$ systemctl stop keepalived kubelet docker
```

Comment:
in shell examples, please remove the `$` characters - https://kubernetes.io/docs/home/contribute/style-guide/#dont-include-the-command-prompt